src/test/java/io/spokestack/spokestack/nlu/tensorflow/TensorflowNLUTest.java

```diff
@@ -0,0 +1,241 @@
+        NLUResult result = env.classify(utterance).get();
+        assertEquals(IllegalStateException.class, result.getError().getClass());
+        assertEquals(utterance, result.getUtterance());
+        assertNull(result.getIntent());
+        assertNull(result.getSlots());
+        assertNull(result.getContext());
+
+        utterance = "this code is for test 1";
+        float[] intentResult =
+              buildIntentResult(2, env.metadata.getIntents().length);
+        float[] tagResult =
+              new float[utterance.split(" ").length
+                    * env.metadata.getTags().length];
+        setTag(tagResult, env.metadata.getTags().length, 0, 1);
+        setTag(tagResult, env.metadata.getTags().length, 1, 2);
+        setTag(tagResult, env.metadata.getTags().length, 5, 3);
+        env.testModel.setOutputs(intentResult, tagResult);
+        result = env.classify(utterance).get();
+
+        Map<String, Slot> slots = new HashMap<>();
+        slots.put("noun_phrase",
+              new Slot("noun_phrase", "this code", "this code"));
+        slots.put("test_num", new Slot("test_num", "1", 1));
+
+        assertNull(result.getError());
+        assertEquals("describe_test", result.getIntent());
+        for (String slotName : slots.keySet()) {
+            assertEquals(slots.get(slotName), result.getSlots().get(slotName));
+        }
+        assertEquals(slots, result.getSlots());
+        assertEquals(utterance, result.getUtterance());
+        assertTrue(result.getContext().isEmpty());
+    }
+
+    private float[] buildIntentResult(int index, int numIntents) {
+        float[] result = new float[numIntents];
+        result[index] = 10;
+        return result;
+    }
+
+    private void setTag(float[] tagResult, int numTags,
+                        int tokenIndex, int tagIndex) {
+        tagResult[tokenIndex * numTags + tagIndex] = 10;
+    }
+
+    public SpeechConfig testConfig() {
+        return new SpeechConfig()
+              .put("nlu-model-path", "model-path")
+              .put("nlu-metadata-path", "src/test/resources/nlu.json")
+              .put("nlu-input-length", 100)
+              .put("wordpiece-vocab-path", "src/test/resources/vocab.txt");
+    }
+
+    public static class TestModel extends TensorflowModel {
+        public TestModel(TensorflowModel.Loader loader) {
+            super(loader);
+        }
+
+        public void run() {
+            this.inputs(0).rewind();
+            this.outputs(0).rewind();
+            this.outputs(1).rewind();
+        }
+
+        public final void setOutputs(float[] intentsOut, float[] tagsOut) {
+            this.outputs(0).rewind();
+            for (float f : intentsOut) {
+                this.outputs(0).putFloat(f);
+            }
+            this.outputs(1).rewind();
+            for (float o : tagsOut) {
+                this.outputs(1).putFloat(o);
+            }
+        }
+    }
+
+    public static class TestEnv implements TextEncoder {
```

Comment:
Impressive committment to mocking here.

Reply:
Not going to lie; I stole it from the existing TensorFlow model tests. I tried to do it a different way, but I didn't need native library wrangling in my testing life.